README.md
# Order Tracker

Make a fabulous order tracker for you e-commerce app is no time.

#### Showcase

![Untitled](https://user-images.githubusercontent.com/102783954/230720601-5ccb5f58-b675-470a-ab4c-154cb22113ad.gif)

### What is different

| Title                                                       | Other Order Tackers | Order Tracker Zen |
| ----------------------------------------------------------- | ------------------- | ----------------- |
| Order Tracking Functionality                                | ✅                  | ✅                |
| Animations                                                  | ✅                  | ✅                |
| Custom Order Headings                                       | 👎🏻                  | ✅                |
| Use Custom Styles                                           | 👎🏻                  | ✅                |
| Toggle bitween expanded and shrinked just with one variable | 👎🏻                  | ✅                |
| Email Support By Zenzzen                                    | 👎🏻                  | ✅                |
| Shrinkable                                                  | 👎🏻                  | ✅                |

**Our team is always happy to help tech enthusiasts like you, you can email us on dev_support@zenzzen.com in case if you find any difficulty**

## How to get started❓ 🤔

You can simply add the dependency from **cli** like:

    flutter pub add order_tracker_zen

**Or** you can add it in your **pubspec.yaml** like:

    dependencies:
    	order_tracker_zen:

## How to use❓ 🤔

Using it is as **easy as** create a **TextFormField**, Don't beleave? Let me show you:

```
OrderTrackerZen(
	tracker_data: [
		TrackerData(
			title: "Order Place",
			date: "Sat, 8 Apr '22",
			tracker_details: [
				TrackerDetails(
					title: "Your order was placed on Zenzzen",
					datetime: "Sat, 8 Apr '22 - 17:17",
				),
				TrackerDetails(
					title: "Zenzzen Arranged A Callback Request",
					datetime: "Sat, 8 Apr '22 - 17:42",
				),
			],
		),
		TrackerData(
			title: "Order Shipped",
			date: "Sat, 8 Apr '22",
			tracker_details: [
				TrackerDetails(
					title: "Your order was shipped with MailDeli",
					datetime: "Sat, 8 Apr '22 - 17:50",
				),
			],
		),
		TrackerData(
			title: "Order Delivered",
			date: "Sat,8 Apr '22",
			tracker_details: [
				TrackerDetails(
					title: "You received your order, by MailDeli",
					datetime: "Sat, 8 Apr '22 - 17:51",
				),
			],
		),
	],
)
```

**Looks Looooong?** Dont worrrrrryyyyy.....
This is because it hardcoded data, in your case you must be using loops here. Ok!

**Let me explain about `OrderTrackerZen`, `TrackerData` and `TrackerDetails` .**

- `OrderTrackerZen` is main container of whole order tracker, like
  <img width="487" alt="image" src="https://user-images.githubusercontent.com/102783954/230720647-97dad7f5-fcb7-4c8f-9f26-e3ac04e94bf6.png">

- `TrackerData` is container of that contains tracking details
  <img width="487" alt="image" src="https://user-images.githubusercontent.com/102783954/230720666-9d501e90-a119-4847-a936-350d0d2afe8d.png">

- `TrackerDetails` is a single tracking tracking detail
  <img width="487" alt="image" src="https://user-images.githubusercontent.com/102783954/230720673-817fd477-aff6-42a5-8398-210da9ecebd6.png">

## Additional Params (Optional)

- `success_color` : Color of progress bar
- `background_color` : Background color of progress bar
- `screen_background_color` : Dot in progress bar has a small border, that must match background, that helps in same
- `text_primary_color` : Primary text color
- `text_secondary_color` : Seconday text color
- `isShrinked` : This boolean toggles bitween shrinked and detailed view

**This is what Zenzzen Team made for you specially 😊, Now it's your responsibility to make ground-breaking app that rocks 🤘🏻**


CHANGELOG.md
# 0.0.0+1

The boom begans

# 0.0.0+2

Feature Updates

# 0.0.0+3

Feature Updates

# 0.0.0+4

Feature Updates

# 0.0.0+5

Feature Updates

# 0.0.0+6

Feature Updates

# 0.0.0+7

Feature Updates

# 0.0.0+8

Feature Updates

# 0.0.0+9

Feature Updates

# 1.0.0+1

Added capabality to shrink

# 1.0.0+2

Opacity of order tracker deatils item shall now change according to tracking progress

# 1.0.0+3

Spacing issue bug solved

# 1.0.0+4

Bug solve


example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility in the flutter_test package. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:order_tracker_example/main.dart';

void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(const MyApp());

    // Verify that our counter starts at 0.
    expect(find.text('0'), findsOneWidget);
    expect(find.text('1'), findsNothing);

    // Tap the '+' icon and trigger a frame.
    await tester.tap(find.byIcon(Icons.add));
    await tester.pump();

    // Verify that our counter has incremented.
    expect(find.text('0'), findsNothing);
    expect(find.text('1'), findsOneWidget);
  });
}


lib/order_tracker_zen.dart
// ignore_for_file: non_constant_identifier_names, prefer_const_constructors, sized_box_for_whitespace, prefer_typing_uninitialized_variables

import 'package:flutter/material.dart';

/// A data model representing a single tracker.
///
/// This model includes a title, date, and list of tracker details.
class TrackerData {
  /// The title of the tracker data.
  final String title;

  /// The date associated with the tracker data.
  final String date;

  /// A list of tracker details.
  final List<TrackerDetails> tracker_details;

  /// Creates a new tracker data instance.
  ///
  /// The [title], [date], and [tracker_details] must not be null.
  TrackerData({
    required this.title,
    required this.date,
    required this.tracker_details,
  });
}

/// A data model representing the details of a tracker.
///
/// This model includes a title and a datetime.
class TrackerDetails {
  /// The title of the tracker details.
  final String title;

  /// The datetime associated with the tracker details.
  final String datetime;

  /// Creates a new tracker details instance.
  ///
  /// The [title] and [datetime] must not be null.
  TrackerDetails({
    required this.title,
    required this.datetime,
  });
}

/// A stateless widget that displays an order tracking progress.
///
/// This widget takes a list of `TrackerData` as input and displays a
/// progress indicator with customizable colors and animation duration.
class OrderTrackerZen extends StatelessWidget {
  /// A list of tracker data.
  final List<TrackerData> tracker_data;

  /// The success color used in the widget.
  final Color? success_color;

  /// The background color used in the widget.
  final Color? background_color;

  /// The screen background color used in the widget.
  final Color? screen_background_color;

  /// The primary text color used in the widget.
  final Color? text_primary_color;

  /// The secondary text color used in the widget.
  final Color? text_secondary_color;

  /// The duration of the animation in milliseconds.
  final int animation_duration;

  /// If the widget is shrinked or not. If true, the widget will be shrinked and will only show title and time.
  final bool isShrinked;

  /// Creates an OrderTrackerZen widget.
  ///
  /// The [tracker_data] must not be null, and the [animation_duration] must be a positive integer.
  const OrderTrackerZen({
    super.key,
    required this.tracker_data,
    this.success_color,
    this.background_color,
    this.screen_background_color,
    this.text_primary_color,
    this.text_secondary_color,
    this.animation_duration = 1500,
    this.isShrinked = false,
  });

  @override
  Widget build(BuildContext context) {
    // Builds the list of tracker items
    return ListView.builder(
      physics: NeverScrollableScrollPhysics(),
      shrinkWrap: true,
      itemBuilder: (context, index) {
        var data = tracker_data[index];
        // Animate the status line
        return TweenAnimationBuilder(
          tween: Tween<double>(
            begin: 0,
            end: tracker_data.length.toDouble(),
          ),
          duration: Duration(
            milliseconds: animation_duration * tracker_data.length,
          ),
          builder: (context, value, child) {
            // Build the main tracker item
            return TweenAnimationBuilder(
              builder: (context, val, child) {
                return val == 1 || val == 0
                    ? OrderTrackerListItem(
                        success_color: success_color,
                        background_color: background_color,
                        screen_background_color: screen_background_color,
                        title: OrderTitleAndDate(
                          isShrinked: isShrinked,
                          title: data.title,
                          date: data.date,
                          text_primary_color: text_primary_color,
                          text_secondary_color: text_secondary_color,
                        ),
                        detailListItems: isShrinked
                            ? []
                            : [
                                for (var i = 0;
                                    i < data.tracker_details.length;
                                    i++)
                                  OrderTrackerDetails(
                                    title: data.tracker_details[i].title,
                                    datetime: data.tracker_details[i].datetime,
                                    text_primary_color: text_primary_color,
                                    text_secondary_color: text_secondary_color,
                                  ),
                              ],
                        isLastItem: index == tracker_data.length - 1,
                        showLine: value > index,
                      )
                    : Container();
              },
              tween: Tween<double>(
                begin: isShrinked ? 0 : 1,
                end: isShrinked ? 1 : 0,
              ),
              duration: Duration(
                milliseconds: 50,
              ),
            );
          },
        );
      },
      itemCount: tracker_data.length,
      padding: EdgeInsets.zero,
    );
  }
}

/// A stateful widget representing a single order tracker list item.
///
/// This widget displays a title, date, and details for a single order tracker item.
class OrderTrackerListItem extends StatefulWidget {
  /// The title and date widget of the main order status widget.
  final OrderTitleAndDate title;

  /// A list of order tracker details.
  final List<OrderTrackerDetails>? detailListItems;

  /// The success color used in the widget.
  final Color? success_color;

  /// The background color used in the widget.
  final Color? background_color;

  /// The screen background color used in the widget.
  final Color? screen_background_color;

  /// This is true if it is the last item in the list.
  final bool isLastItem;

  // This is true if the status line should be shown.
  final bool showLine;

  /// Creates an OrderTrackerListItem widget.
  ///
  /// The [title] must not be null, [isLastItem] and [showLine] default to false.
  const OrderTrackerListItem({
    super.key,
    required this.title,
    this.detailListItems,
    this.isLastItem = false,
    this.showLine = false,
    this.success_color,
    this.background_color,
    this.screen_background_color,
  });

  @override
  State<OrderTrackerListItem> createState() => _OrderTrackerListItemState();
}

class _OrderTrackerListItemState extends State<OrderTrackerListItem> {
  /// The key of the row widget. Used to get the height of the row.
  GlobalKey rowKey = GlobalKey();

  /// Gets the height of the row.
  double getRowHeight() {
    final RenderBox? rowRenderBox =
        rowKey.currentContext?.findRenderObject() as RenderBox?;
    return rowRenderBox?.size.height ?? 0.0;
  }

  // This is the height of the row, it shall be initialized later in the build method.
  var rowHeight = 0.0;

  @override
  void initState() {
    super.initState();
    WidgetsBinding.instance.addPostFrameCallback((_) {
      setState(() {
        rowHeight = getRowHeight();
      });
    });
  }

  @override
  Widget build(BuildContext context) {
    return Stack(
      children: [
        widget.isLastItem
            ? Container()
            : Padding(
                padding: EdgeInsets.only(left: 5),
                child: Container(
                  width: 2,
                  height: rowHeight,
                  child: RotatedBox(
                    quarterTurns: 1,
                    child: TweenAnimationBuilder(
                      builder: (context, value, child) {
                        return LinearProgressIndicator(
                          value: value,
                          color: widget.success_color ?? Colors.green,
                          backgroundColor:
                              widget.background_color ?? Colors.grey[400],
                        );
                      },
                      tween:
                          Tween<double>(begin: 0, end: widget.showLine ? 1 : 0),
                      duration: Duration(milliseconds: 1500),
                    ),
                  ),
                ),
              ),
        Row(
          key: rowKey,
          crossAxisAlignment: CrossAxisAlignment.start,
          children: [
            // Dot widget
            Container(
              width: 12,
              height: 12,
              decoration: BoxDecoration(
                color: widget.success_color ?? Colors.green,
                shape: BoxShape.circle,
                border: Border.all(
                  color: widget.screen_background_color ?? Colors.white,
                  width: 2,
                ),
              ),
            ),
            SizedBox(width: 10),
            // Title and date widget
            Expanded(
              child: TweenAnimationBuilder(
                builder: (context, value, child) {
                  return Opacity(
                    opacity: value,
                    child: Column(
                      crossAxisAlignment: CrossAxisAlignment.start,
                      children: [
                        widget.title,
                        widget.detailListItems == null
                            ? SizedBox()
                            : Column(
                                crossAxisAlignment: CrossAxisAlignment.start,
                                children: widget.detailListItems!,
                              ),
                        widget.detailListItems == null
                            ? SizedBox()
                            : SizedBox(height: widget.isLastItem ? 0 : 13),
                      ],
                    ),
                  );
                },
                tween: Tween<double>(
                  begin: 0.4,
                  end: widget.showLine ? 1 : 0.4,
                ),
                duration: Duration(milliseconds: 500),
              ),
            ),
          ],
        ),
      ],
    );
  }
}

/// A stateless widget representing the details of an order tracker.
///
/// This widget displays a title and a datetime.
class OrderTrackerDetails extends StatelessWidget {
  /// The title of the order tracker details.
  final title;

  /// The datetime associated with the order tracker details.
  final datetime;

  /// The primary text color used in the widget.
  final Color? text_primary_color;

  /// The secondary text color used in the widget.
  final Color? text_secondary_color;

  /// Creates an OrderTrackerDetails widget.
  ///
  /// The [title] and [datetime] must not be null.
  const OrderTrackerDetails({
    super.key,
    required this.title,
    required this.datetime,
    this.text_primary_color,
    this.text_secondary_color,
  });

  @override
  Widget build(BuildContext context) {
    return Padding(
      padding: const EdgeInsets.all(8.0),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          Text(
            title,
            style: TextStyle(
              color: text_primary_color ?? Colors.black,
              fontSize: 12,
            ),
          ),
          SizedBox(height: 5),
          Text(
            datetime,
            style: TextStyle(
              color: text_secondary_color ?? Colors.grey,
              fontSize: 11,
            ),
          ),
        ],
      ),
    );
  }
}

/// A stateless widget representing the title and date of an order tracker.
///
/// This widget displays a title and a date.
class OrderTitleAndDate extends StatelessWidget {
  /// The title of the order tracker.
  final title;

  /// The date associated with the order tracker.
  final date;

  /// The primary text color used in the widget.
  final Color? text_primary_color;

  /// The secondary text color used in the widget.
  final Color? text_secondary_color;

  /// If the widget is shrinked or not. If true, the widget will be shrinked and will only show title and time.
  final bool isShrinked;

  /// Creates an OrderTitleAndDate widget.
  ///
  /// The [title] and [date] must not be null.
  const OrderTitleAndDate({
    super.key,
    required this.title,
    required this.date,
    this.text_primary_color,
    this.text_secondary_color,
    this.isShrinked = false,
  });

  @override
  Widget build(BuildContext context) {
    var children = <Widget>[
      Text(
        title,
        style: TextStyle(
          color: text_primary_color ?? Colors.black,
          fontSize: 15,
          fontWeight: FontWeight.bold,
        ),
      ),
      SizedBox(width: isShrinked ? 0 : 5, height: isShrinked ? 5 : 0),
      Text(
        date,
        style: TextStyle(
          color: text_secondary_color ?? Colors.grey,
          fontSize: isShrinked ? 13 : 15,
          fontWeight: FontWeight.normal,
        ),
      ),
    ];
    return Padding(
      padding: EdgeInsets.only(
          left: 8, right: 8, bottom: isShrinked ? 0 : 8, top: 0),
      child: isShrinked
          ? Column(
              crossAxisAlignment: CrossAxisAlignment.start,
              children: children,
            )
          : Row(
              crossAxisAlignment: CrossAxisAlignment.center,
              children: children,
            ),
    );
  }
}


example/linux/my_application.cc
#include "my_application.h"

#include <flutter_linux/flutter_linux.h>
#ifdef GDK_WINDOWING_X11
#include <gdk/gdkx.h>
#endif

#include "flutter/generated_plugin_registrant.h"

struct _MyApplication {
  GtkApplication parent_instance;
  char** dart_entrypoint_arguments;
};

G_DEFINE_TYPE(MyApplication, my_application, GTK_TYPE_APPLICATION)

// Implements GApplication::activate.
static void my_application_activate(GApplication* application) {
  MyApplication* self = MY_APPLICATION(application);
  GtkWindow* window =
      GTK_WINDOW(gtk_application_window_new(GTK_APPLICATION(application)));

  // Use a header bar when running in GNOME as this is the common style used
  // by applications and is the setup most users will be using (e.g. Ubuntu
  // desktop).
  // If running on X and not using GNOME then just use a traditional title bar
  // in case the window manager does more exotic layout, e.g. tiling.
  // If running on Wayland assume the header bar will work (may need changing
  // if future cases occur).
  gboolean use_header_bar = TRUE;
#ifdef GDK_WINDOWING_X11
  GdkScreen* screen = gtk_window_get_screen(window);
  if (GDK_IS_X11_SCREEN(screen)) {
    const gchar* wm_name = gdk_x11_screen_get_window_manager_name(screen);
    if (g_strcmp0(wm_name, "GNOME Shell") != 0) {
      use_header_bar = FALSE;
    }
  }
#endif
  if (use_header_bar) {
    GtkHeaderBar* header_bar = GTK_HEADER_BAR(gtk_header_bar_new());
    gtk_widget_show(GTK_WIDGET(header_bar));
    gtk_header_bar_set_title(header_bar, "order_tracker_example");
    gtk_header_bar_set_show_close_button(header_bar, TRUE);
    gtk_window_set_titlebar(window, GTK_WIDGET(header_bar));
  } else {
    gtk_window_set_title(window, "order_tracker_example");
  }

  gtk_window_set_default_size(window, 1280, 720);
  gtk_widget_show(GTK_WIDGET(window));

  g_autoptr(FlDartProject) project = fl_dart_project_new();
  fl_dart_project_set_dart_entrypoint_arguments(project, self->dart_entrypoint_arguments);

  FlView* view = fl_view_new(project);
  gtk_widget_show(GTK_WIDGET(view));
  gtk_container_add(GTK_CONTAINER(window), GTK_WIDGET(view));

  fl_register_plugins(FL_PLUGIN_REGISTRY(view));

  gtk_widget_grab_focus(GTK_WIDGET(view));
}

// Implements GApplication::local_command_line.
static gboolean my_application_local_command_line(GApplication* application, gchar*** arguments, int* exit_status) {
  MyApplication* self = MY_APPLICATION(application);
  // Strip out the first argument as it is the binary name.
  self->dart_entrypoint_arguments = g_strdupv(*arguments + 1);

  g_autoptr(GError) error = nullptr;
  if (!g_application_register(application, nullptr, &error)) {
     g_warning("Failed to register: %s", error->message);
     *exit_status = 1;
     return TRUE;
  }

  g_application_activate(application);
  *exit_status = 0;

  return TRUE;
}

// Implements GObject::dispose.
static void my_application_dispose(GObject* object) {
  MyApplication* self = MY_APPLICATION(object);
  g_clear_pointer(&self->dart_entrypoint_arguments, g_strfreev);
  G_OBJECT_CLASS(my_application_parent_class)->dispose(object);
}

static void my_application_class_init(MyApplicationClass* klass) {
  G_APPLICATION_CLASS(klass)->activate = my_application_activate;
  G_APPLICATION_CLASS(klass)->local_command_line = my_application_local_command_line;
  G_OBJECT_CLASS(klass)->dispose = my_application_dispose;
}

static void my_application_init(MyApplication* self) {}

MyApplication* my_application_new() {
  return MY_APPLICATION(g_object_new(my_application_get_type(),
                                     "application-id", APPLICATION_ID,
                                     "flags", G_APPLICATION_NON_UNIQUE,
                                     nullptr));
}


example/windows/runner/main.cpp
#include <flutter/dart_project.h>
#include <flutter/flutter_view_controller.h>
#include <windows.h>

#include "flutter_window.h"
#include "utils.h"

int APIENTRY wWinMain(_In_ HINSTANCE instance, _In_opt_ HINSTANCE prev,
                      _In_ wchar_t *command_line, _In_ int show_command) {
  // Attach to console when present (e.g., 'flutter run') or create a
  // new console when running with a debugger.
  if (!::AttachConsole(ATTACH_PARENT_PROCESS) && ::IsDebuggerPresent()) {
    CreateAndAttachConsole();
  }

  // Initialize COM, so that it is available for use in the library and/or
  // plugins.
  ::CoInitializeEx(nullptr, COINIT_APARTMENTTHREADED);

  flutter::DartProject project(L"data");

  std::vector<std::string> command_line_arguments =
      GetCommandLineArguments();

  project.set_dart_entrypoint_arguments(std::move(command_line_arguments));

  FlutterWindow window(project);
  Win32Window::Point origin(10, 10);
  Win32Window::Size size(1280, 720);
  if (!window.Create(L"order_tracker_example", origin, size)) {
    return EXIT_FAILURE;
  }
  window.SetQuitOnClose(true);

  ::MSG msg;
  while (::GetMessage(&msg, nullptr, 0, 0)) {
    ::TranslateMessage(&msg);
    ::DispatchMessage(&msg);
  }

  ::CoUninitialize();
  return EXIT_SUCCESS;
}


example/lib/main.dart
/// Order Tracker Zen
///
/// A Flutter package that provides a simple and customizable order tracking widget for your applications.
/// This example demonstrates how to create an order tracking widget using the OrderTrackerZen package.
///
/// To use this package, add `order_tracker_zen` as a dependency in your `pubspec.yaml` file.
import 'package:flutter/material.dart';
import 'package:order_tracker_zen/order_tracker_zen.dart';

/// The main function is the entry point of the application.
void main(List<String> args) {
  runApp(MyApp());
}

/// MyApp is a StatelessWidget that acts as the root widget of the application.
///
/// It configures the MaterialApp with the necessary theme and routing information.
class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      theme: ThemeData(
        useMaterial3: true,
      ),
      home: Scaffold(
        appBar: AppBar(
          title: const Text("Order Tracker Zen"),
        ),
        body: Center(
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.center,
            children: [
              // Add padding around the OrderTrackerZen widget for better presentation.
              Padding(
                padding: EdgeInsets.symmetric(horizontal: 20, vertical: 30),
                // OrderTrackerZen is the main widget of the package which displays the order tracking information.
                child: OrderTrackerZen(
                  // Provide an array of TrackerData objects to display the order tracking information.
                  tracker_data: [
                    // TrackerData represents a single step in the order tracking process.
                    TrackerData(
                      title: "Order Placed",
                      date: "Sat, 8 Apr '22",
                      // Provide an array of TrackerDetails objects to display more details about this step.
                      tracker_details: [
                        // TrackerDetails contains detailed information about a specific event in the order tracking process.
                        TrackerDetails(
                          title: "Your order was placed on Zenzzen",
                          datetime: "Sat, 8 Apr '22 - 17:17",
                        ),
                        TrackerDetails(
                          title: "Zenzzen Arranged A Callback Request",
                          datetime: "Sat, 8 Apr '22 - 17:42",
                        ),
                      ],
                    ),
                    // yet another TrackerData object
                    TrackerData(
                      title: "Order Shipped",
                      date: "Sat, 8 Apr '22",
                      tracker_details: [
                        TrackerDetails(
                          title: "Your order was shipped with MailDeli",
                          datetime: "Sat, 8 Apr '22 - 17:50",
                        ),
                      ],
                    ),
                    // And yet another TrackerData object
                    TrackerData(
                      title: "Order Delivered",
                      date: "Sat,8 Apr '22",
                      tracker_details: [
                        TrackerDetails(
                          title: "You received your order, by MailDeli",
                          datetime: "Sat, 8 Apr '22 - 17:51",
                        ),
                      ],
                    ),
                  ],
                ),
              ),
            ],
          ),
        ),
      ),
    );
  }
}

// Path: example/lib/main.dart
// Order Tracker Zen
// A Flutter package that provides a simple and customizable order tracking widget for your applications.
// This example demonstrates how to create an order tracking widget using the OrderTrackerZen package.
// To use this package, add `order_tracker_zen` as a dependency in your `pubspec.yaml` file.
